File: messaging/grpc/explainer/node/explain.js
var PROTO_PATH = __dirname + '/../protocols/explainer.proto';

var grpc = require('grpc');
var explainer_proto = grpc.load(PROTO_PATH).explainer;

function main() {
  var client = new explainer_proto.ShiFu('localhost:50051', grpc.credentials.createInsecure());
  var problem = 'help';
  if (process.argv.length >= 3) {
    problem = process.argv[2];
  }
  client.tellMeWhy({problem: problem}, function(err, response) {
    console.log(response.explanation);
  });
}

main();

File: messaging/grpc/explainer/cpp/explain.cc
#include <iostream>
#include <memory>
#include <string>

#include <grpc++/grpc++.h>

#include "explainer.grpc.pb.h"

using grpc::Channel;
using grpc::ClientContext;
using grpc::Status;
using explainer::ExplainerRequest;
using explainer::ExplainerReply;
using explainer::ShiFu;

class ExplainerClient {
  public:
    ExplainerClient(std::shared_ptr<Channel> channel)
      : stub_(ShiFu::NewStub(channel)) {}

  std::string TellMeWhy(const std::string& problem) {
    ExplainerRequest request;
    request.set_problem(problem);

    ExplainerReply reply;

    ClientContext context;

    Status status = stub_->TellMeWhy(&context, request, &reply);

    if (status.ok()) {
      return reply.explanation();
    } else {
      return "RPC failed";
    }
  }

 private:
  std::unique_ptr<ShiFu::Stub> stub_;
};

int main(int argc, char** argv) {
  ExplainerClient client(grpc::CreateChannel("localhost:50051", grpc::InsecureChannelCredentials()));
  std::string problem;
  if(argc>1) {
    problem = argv[1];
  } else {
    problem = "help";
  }
  std::string reply = client.TellMeWhy(problem);
  std::cout << reply << std::endl;
  return 0;
}
